run/src/main/kotlin/com/cosmotech/run/service/RunServiceImpl.kt

```diff
@@ -0,0 +1,224 @@
+import com.cosmotech.runner.service.getRbac
+import java.time.Instant
+import org.springframework.context.event.EventListener
+import org.springframework.data.domain.PageRequest
+import org.springframework.stereotype.Service
+
+internal const val WORKFLOW_TYPE_RUN = "container-run"
+
+@Service
+@Suppress("TooManyFunctions")
+class RunServiceImpl(
+    private val containerFactory: RunContainerFactory,
+    private val workflowService: WorkflowService,
+    private val runnerApiService: RunnerApiServiceInterface,
+    private val runRepository: RunRepository,
+    private val csmRbac: CsmRbac
+) : CsmPhoenixService(), RunApiServiceInterface {
+
+  override fun listRuns(
+      organizationId: String,
+      workspaceId: String,
+      runnerId: String,
+      page: Int?,
+      size: Int?
+  ): List<Run> {
+    // This call verify by itself that we have the read authorization in the scenario
+    val runner = runnerApiService.getRunner(organizationId, workspaceId, runnerId)
+    csmRbac.verify(runner.getRbac(), PERMISSION_READ, getScenarioRolesDefinition())
+
+    val defaultPageSize = csmPlatformProperties.twincache.scenariorun.defaultPageSize
+    val pageable =
+        constructPageRequest(page, size, defaultPageSize) ?: PageRequest.of(0, defaultPageSize)
+
+    // We know which page to return
+    return runRepository
+        .findByRunnerId(organizationId, workspaceId, runnerId, pageable)
+        .toList()
+        .map { it.withStateInformation().withoutSensitiveData()!! }
+  }
+
+  private fun Run?.withStateInformation(): Run? {
+    if (this?.state?.isTerminal() == true || this == null) {
+      return this
+    }
+
+    val state = getRunStatus(this).state
+    val run = this.copy(state = state)
+
+    if (state?.isTerminal() == true) {
+      runRepository.save(run)
+    }
+    return run
+  }
+
+  private fun getRunStatus(run: Run): RunStatus {
+    val runStatus = this.workflowService.getRunStatus(run)
+    return runStatus.copy(
+        state = mapWorkflowPhaseToRunStatus(phase = runStatus.phase, runId = run.id))
+  }
+
+  private fun mapWorkflowPhaseToRunStatus(phase: String?, runId: String?): RunState {
+    return when (phase) {
+      "Pending",
+      "Running" -> RunState.Running
+      "Succeeded" -> RunState.Successful
+      "Skipped",
+      "Failed",
+      "Error",
+      "Omitted" -> RunState.Failed
+      else -> {
+        logger.warn("Unhandled state response for job {}: {} => Returning Unknown", runId, phase)
+        RunState.Unknown
+      }
+    }
+  }
+
+  override fun getRun(
+      organizationId: String,
+      workspaceId: String,
+      runnerId: String,
+      runId: String
+  ): Run {
+    val run =
+        runRepository
+            .findBy(organizationId, workspaceId, runnerId, runId)
+            .orElseGet { null }
+            .withStateInformation()
+            .withoutSensitiveData()
+            ?: throw java.lang.IllegalArgumentException("Run #$runId not found in #$runnerId")
```

Comment:
You can remove `java.lang`
Plus it could be rewrite like bellow:

```
    runRepository
    .findBy(organizationId, workspaceId, runnerId, runId)
    .orElseThrow{
        throw IllegalArgumentException("Run #$runId not found in #$runnerId")
    }
    .withStateInformation()
    .withoutSensitiveData()

Reply:
rewrite and remove the run? from withStateInformation() function and withoutSensitiveData() function